src/wallet/wallet_tx_builder.cpp

```diff
@@ -0,0 +1,751 @@
+                if (acct.GetAccountId() == ZCASH_LEGACY_ACCOUNT) {
+                    tfvk = pwalletMain->GetLegacyAccountKey().ToAccountPubKey();
+                } else {
+                    // By definition, we have a UFVK for every known account.
+                    auto ufvk = pwalletMain->GetUnifiedFullViewingKeyByAccount(acct.GetAccountId()).value();
+                    // Transparent UTXOs will not have been selected if the UFVK does not contain
+                    // a transparent key.
+                    tfvk = ufvk.GetTransparentKey().value();
+                }
+            },
+            [&](const auto& other) {
+                throw std::runtime_error("SelectOVKs: Selector cannot select transparent UTXOs.");
+            }
+        }, selector.GetPattern());
+        assert(tfvk.has_value());
+
+        auto ovks = tfvk.value().GetOVKsForShielding();
+        internalOVK = ovks.first;
+        externalOVK = ovks.second;
+    } else if (!spendable.sproutNoteEntries.empty()) {
+        // use the legacy transparent account OVKs when sending from Sprout
+        auto tfvk = pwalletMain->GetLegacyAccountKey().ToAccountPubKey();
+        auto ovks = tfvk.GetOVKsForShielding();
+        internalOVK = ovks.first;
+        externalOVK = ovks.second;
+    } else {
+        // This should be unreachable; it is left in place as a guard to ensure
+        // that when new input types are added to SpendableInputs in the future
+        // that we do not accidentally return the all-zeros OVK.
+        throw std::runtime_error("No spendable inputs.");
+    }
+
+    return std::make_pair(internalOVK, externalOVK);
+}
+
+PrivacyPolicy TransactionEffects::GetRequiredPrivacyPolicy() const
+{
+    PrivacyPolicy maxPrivacy = PrivacyPolicy::FullPrivacy;
+
+    if (!spendable.orchardNoteMetadata.empty() && payments.HasSaplingRecipient()) {
+        maxPrivacy = PrivacyPolicy::AllowRevealedAmounts;
+    }
+
+    if (!spendable.saplingNoteEntries.empty() && payments.HasOrchardRecipient()) {
+        maxPrivacy = PrivacyPolicy::AllowRevealedAmounts;
+    }
+
+    if (!spendable.sproutNoteEntries.empty() && payments.HasSaplingRecipient()) {
+        maxPrivacy = PrivacyPolicy::AllowRevealedAmounts;
+    }
+
+    bool hasTransparentSource = !spendable.utxos.empty();
+    if (payments.HasTransparentRecipient()) {
+        if (hasTransparentSource) {
+            // TODO: This is the correct policy, but it’s a breaking change from previous behavior,
+            //       so enable it separately.
+            // maxPrivacy = PrivacyPolicy::AllowFullyTransparent;
+            maxPrivacy = PrivacyPolicy::AllowRevealedSenders;
+        } else {
+            maxPrivacy = PrivacyPolicy::AllowRevealedRecipients;
+        }
+    } else if (hasTransparentSource) {
+        maxPrivacy = PrivacyPolicy::AllowRevealedSenders;
+    }
+
+    // TODO: Check for conditions where PrivacyPolicy::AllowLinkingAccountAccesses
+    // or PrivacyPolicy::NoPrivacy are required
```

Comment:
Similar question about this TODO: is this coming later?

Reply:
Sort of … there’s a PR I’m about to push that uses the `PrivacyPolicyMeet` work from #6252 which obviates the need to check for `NoPrivacy` (as it’s only ever required as the meet of other requirements). However, that doesn’t cover `AllowLinkingAccountAddresses`, which should be checked in this PR.

Is the precise requirement for `AllowLinkingAccountAddresses` that there are transparent components spent from multiple UAs?

More generally, I was thinking that we could (in a future PR) eliminate the duplicate `PrivacyPolicy` checking (first, scattered in the code, as we’re preparing the tx; then second, here, against the finished tx). Instead, we build the tx optimistically, then only check against the policy here. Probably this will require some other changes where we currently use the PrivacyPolicy to inform the tx instead of fail (e.g., ignoring transparent inputs because the tx doesn’t `AllowRevealedSenders`). But even leaving those cases in for now, it would simplify the policy checks and get us closer to the 2-stage approach where there isn’t an explicit PrivacyPolicy.

Reply:
This has been implemented in this PR now.